appinventor/blocklyeditor/src/versioning.js

```diff
@@ -40,6 +40,32 @@ Blockly.Versioning.log = function log(string) { // Display feedback on upgrade i
   }
 };
 
+/**
+ * Updates the `<mutation>` element of `component_method` blocks to include a `shape` field that
+ * the block can use to shape itself properly in the event that the method definition is
+ * unavailable.
+ *
+ * @param {Element} dom the root of the workspace XML document
+ * @returns {Element}
+ */
+Blockly.Versioning.upgradeComponentMethods = function(dom) {
+  var els = dom.querySelectorAll('block[type="component_method"]');
+  els.forEach(function(el) {
+    var parent = el.parentElement;
+    var mutation = el.querySelector('mutation');
+    if (!mutation) {
+      console.warn('component_method without mutation', el);
+      return;
+    }
+    if (!parent) {
+      // detached block
+      return;
+    }
+    mutation.setAttribute('shape', parent.tagName.toLowerCase());
```

Comment:
As a heads up: In JSON we no longer store the type of the input (because it doesn't actually affect serialization). So if/when you transition to the JSON system, you'll probably (?) need to include this via `mutationToDom`.

Example of JSON state:
```
{
  "blocks": {
    "languageVersion": 0,
    "blocks": [
      {
        "type": "controls_if",
        "inputs": {
          "IF0": {
            "block": {
              "type": "logic_boolean",
              "fields": {
                "BOOL": "TRUE"
              }
            }
          },
          "DO0": {
            "block": {
              "type": "text_print",
            }
          }
        }
      }
    ]
  }
}
```

Reply:
Yes. I imagine when we upgrade everything over to JSON we'll want to make sure to update the block definition accordingly. Of course, we will still will need to keep this fix around to address older workspaces with this issue.